# Working with real data
Data from https://www.kaggle.com/nasa/astronaut-yearbook

In [1]:
import pandas as pd

## Dataframe
A dataframe is the most versatile data structure in pandas. You can think of it as an excel sheet with columns and rows.

## Import data from CSV file

In [2]:
df = pd.read_csv("./astronauts.csv")

## Dataframe properties

You can get an overview of the dataframe values with

In [3]:
df.describe()

Year       Group  Space Flights  Space Flight (hr)  Space Walks  \
count   330.000000  330.000000     357.000000         357.000000   357.000000   
mean   1985.106061   11.409091       2.364146        1249.266106     1.246499   
std      13.216147    5.149962       1.428700        1896.759857     2.056989   
min    1959.000000    1.000000       0.000000           0.000000     0.000000   
25%    1978.000000    8.000000       1.000000         289.000000     0.000000   
50%    1987.000000   12.000000       2.000000         590.000000     0.000000   
75%    1996.000000   16.000000       3.000000        1045.000000     2.000000   
max    2009.000000   20.000000       7.000000       12818.000000    10.000000   

       Space Walks (hr)  
count        357.000000  
mean           7.707283  
std           13.367973  
min            0.000000  
25%            0.000000  
50%            0.000000  
75%           12.000000  
max           67.000000

With the len function You can get the number of rows in the dataset

In [4]:
len(df)

357

which gives us 357 astronauts

The columns property gives you the names of the individual columns

In [5]:
df.columns


Index(['Name', 'Year', 'Group', 'Status', 'Birth Date', 'Birth Place',
       'Gender', 'Alma Mater', 'Undergraduate Major', 'Graduate Major',
       'Military Rank', 'Military Branch', 'Space Flights',
       'Space Flight (hr)', 'Space Walks', 'Space Walks (hr)', 'Missions',
       'Death Date', 'Death Mission'],
      dtype='object')

The methods head() gives you the first n (default=5) entries:

In [6]:
df.head()


Name    Year  Group   Status Birth Date    Birth Place Gender  \
0   Joseph M. Acaba  2004.0   19.0   Active  5/17/1967  Inglewood, CA   Male   
1    Loren W. Acton     NaN    NaN  Retired   3/7/1936   Lewiston, MT   Male   
2  James C. Adamson  1984.0   10.0  Retired   3/3/1946     Warsaw, NY   Male   
3   Thomas D. Akers  1987.0   12.0  Retired  5/20/1951  St. Louis, MO   Male   
4       Buzz Aldrin  1963.0    3.0  Retired  1/20/1930  Montclair, NJ   Male   

                                          Alma Mater     Undergraduate Major  \
0  University of California-Santa Barbara; Univer...                 Geology   
1   Montana State University; University of Colorado     Engineering Physics   
2          US Military Academy; Princeton University             Engineering   
3                       University of Missouri-Rolla     Applied Mathematics   
4                           US Military Academy; MIT  Mechanical Engineering   

          Graduate Major Military Rank         Military Branch  Space Flights  \
0                Geology           NaN                     NaN              2   
1          Solar Physics           NaN                     NaN              1   
2  Aerospace Engineering       Colonel       US Army (Retired)              2   
3    Applied Mathematics       Colonel  US Air Force (Retired)              4   
4           Astronautics       Colonel  US Air Force (Retired)              2   

   Space Flight (hr)  Space Walks  Space Walks (hr)  \
0               3307            2              13.0   
1                190            0               0.0   
2                334            0               0.0   
3                814            4              29.0   
4                289            2               8.0   

                                            Missions Death Date Death Mission  
0             STS-119 (Discovery), ISS-31/32 (Soyuz)        NaN           NaN  
1                              STS 51-F (Challenger)        NaN           NaN  
2               STS-28 (Columbia), STS-43 (Atlantis)        NaN           NaN  
3  STS-41 (Discovery), STS-49 (Endeavor), STS-61 ...        NaN           NaN  
4                               Gemini 12, Apollo 11        NaN           NaN

whereas the tail method gives you the last n entries

In [7]:
df.tail(7)


Name    Year  Group   Status  Birth Date  \
350        G. Reid Wiseman  2009.0   20.0   Active  11/11/1975   
351     Peter J. K. Wisoff  1990.0   13.0  Retired   8/16/1958   
352          David A. Wolf  1990.0   13.0  Retired   8/23/1956   
353  Neil W. Woodward III   1998.0   17.0  Retired   7/26/1962   
354       Alfred M. Worden  1966.0    5.0  Retired    2/7/1932   
355          John W. Young  1962.0    2.0  Retired   9/24/1930   
356        George D. Zamka  1998.0   17.0  Retired   6/29/1962   

           Birth Place Gender  \
350      Baltimore, MD   Male   
351        Norfolk, VA   Male   
352   Indianapolis, IN   Male   
353        Chicago, IL   Male   
354        Jackson, MI   Male   
355  San Francisco, CA   Male   
356    Jersey City, NJ   Male   

                                            Alma Mater  \
350  Rensselaer Polytechnic Institute; Johns Hopkin...   
351        University of Virginia; Stanford University   
352              Purdue University; Indiana University   
353  MIT; University of Texas-Austin; George Washin...   
354        US Military Academy; University of Michigan   
355                    Georgia Institute of Technology   
356  US Naval Academy; Florida Institute of Technology   

                Undergraduate Major                            Graduate Major  \
350  Computer & Systems Engineering                       Systems Engineering   
351                         Physics                           Applied Physics   
352          Electrical Engineering                                  Medicine   
353                         Physics              Physics; Business Management   
354                Military Science  Aeronautical & Astronautical Engineering   
355        Aeronautical Engineering                                       NaN   
356                     Mathematics                    Engineering Management   

    Military Rank            Military Branch  Space Flights  \
350     Commander                    US Navy              1   
351           NaN                        NaN              4   
352           NaN                        NaN              3   
353     Commander                    US Navy              0   
354       Colonel     US Air Force (Retired)              1   
355       Captain          US Navy (Retired)              6   
356       Colonel  US Marine Corps (Retired)              2   

     Space Flight (hr)  Space Walks  Space Walks (hr)  \
350               3968            2              13.0   
351               1064            3              20.0   
352               4044            7              41.0   
353                  0            0               0.0   
354                295            1               0.5   
355                835            3              20.0   
356                692            0               0.0   

                                              Missions Death Date  \
350                                  ISS-40/41 (Soyuz)        NaN   
351  STS-57 (Endeavor), STS-68 (Endeavor), STS-81 (...        NaN   
352  STS-58 (Columbia). STS-86/89 (Atlantis/Endeavo...        NaN   
353                                                NaN        NaN   
354                                          Apollo 15        NaN   
355  Gemini 3, Gemini 10, Apollo 10, Apollo 16, STS...        NaN   
356            STS-120 (Discovery), STS-130 (Endeavor)        NaN   

    Death Mission  
350           NaN  
351           NaN  
352           NaN  
353           NaN  
354           NaN  
355           NaN  
356           NaN

With the iloc keyword You get the entries directly via index

In [8]:
df.iloc[0, 0]


'Joseph M. Acaba'

The loc keyword is another way to access dataframe.

The colon is used as a "select *" for rows or columns

In [9]:
df.iloc[:, :]


Name    Year  Group   Status Birth Date  \
0          Joseph M. Acaba  2004.0   19.0   Active  5/17/1967   
1           Loren W. Acton     NaN    NaN  Retired   3/7/1936   
2         James C. Adamson  1984.0   10.0  Retired   3/3/1946   
3          Thomas D. Akers  1987.0   12.0  Retired  5/20/1951   
4              Buzz Aldrin  1963.0    3.0  Retired  1/20/1930   
..                     ...     ...    ...      ...        ...   
352          David A. Wolf  1990.0   13.0  Retired  8/23/1956   
353  Neil W. Woodward III   1998.0   17.0  Retired  7/26/1962   
354       Alfred M. Worden  1966.0    5.0  Retired   2/7/1932   
355          John W. Young  1962.0    2.0  Retired  9/24/1930   
356        George D. Zamka  1998.0   17.0  Retired  6/29/1962   

           Birth Place Gender  \
0        Inglewood, CA   Male   
1         Lewiston, MT   Male   
2           Warsaw, NY   Male   
3        St. Louis, MO   Male   
4        Montclair, NJ   Male   
..                 ...    ...   
352   Indianapolis, IN   Male   
353        Chicago, IL   Male   
354        Jackson, MI   Male   
355  San Francisco, CA   Male   
356    Jersey City, NJ   Male   

                                            Alma Mater  \
0    University of California-Santa Barbara; Univer...   
1     Montana State University; University of Colorado   
2            US Military Academy; Princeton University   
3                         University of Missouri-Rolla   
4                             US Military Academy; MIT   
..                                                 ...   
352              Purdue University; Indiana University   
353  MIT; University of Texas-Austin; George Washin...   
354        US Military Academy; University of Michigan   
355                    Georgia Institute of Technology   
356  US Naval Academy; Florida Institute of Technology   

          Undergraduate Major                            Graduate Major  \
0                     Geology                                   Geology   
1         Engineering Physics                             Solar Physics   
2                 Engineering                     Aerospace Engineering   
3         Applied Mathematics                       Applied Mathematics   
4      Mechanical Engineering                              Astronautics   
..                        ...                                       ...   
352    Electrical Engineering                                  Medicine   
353                   Physics              Physics; Business Management   
354          Military Science  Aeronautical & Astronautical Engineering   
355  Aeronautical Engineering                                       NaN   
356               Mathematics                    Engineering Management   

    Military Rank            Military Branch  Space Flights  \
0             NaN                        NaN              2   
1             NaN                        NaN              1   
2         Colonel          US Army (Retired)              2   
3         Colonel     US Air Force (Retired)              4   
4         Colonel     US Air Force (Retired)              2   
..            ...                        ...            ...   
352           NaN                        NaN              3   
353     Commander                    US Navy              0   
354       Colonel     US Air Force (Retired)              1   
355       Captain          US Navy (Retired)              6   
356       Colonel  US Marine Corps (Retired)              2   

     Space Flight (hr)  Space Walks  Space Walks (hr)  \
0                 3307            2              13.0   
1                  190            0               0.0   
2                  334            0               0.0   
3                  814            4              29.0   
4                  289            2               8.0   
..                 ...          ...               ...   
352               4044            7              41.0   
353                  0           

### Which American astronauts have spent the most time in space? 

In [10]:
most_time_in_space = df.sort_values(by="Space Flight (hr)", ascending=False).head(1)


In [11]:
most_time_in_space[['Name','Space Flight (hr)']]

Name  Space Flight (hr)
346  Jeffrey N. Williams              12818

Sorting the dataframe can be done with sort_by_values. And for this question we sort for Space Flight (hr). Because we want the most hours we have to sort descending which translates to ascending=False.

head(1) gives us the correct answer:

Jeffrey N. Williams. He spent 12818 hours (534 days) in space.

Have You heard of him? Unsung hero!

Hint: the Dataset was updated the last time in 2017. As of 2019 Peggy Whitson is the american who has spent the most time in space. 

She has spend more than 665 days in space!

In [12]:
least_time_in_space = df.sort_values(by="Space Flight (hr)", ascending=True).head(10)
least_time_in_space[['Name','Space Flight (hr)']]

Name  Space Flight (hr)
318    Stephen D. Thorne                  0
246     Brian T. O'Leary                  0
326    Mark T. Vande Hei                  0
173    Gregory B. Jarvis                  0
46          John S. Bull                  0
322      Scott D. Tingle                  0
165  Donald L. Holmquest                  0
50       Yvonne D. Cagle                  0
51     Fernando Caldeiro                  0
63      Roger B. Chaffee                  0

### What university has produced the most astronauts?

In [14]:
df['Alma Mater'].value_counts().head(3)

Alma Mater
US Naval Academy                                  12
US Naval Academy; US Naval Postgraduate School    11
US Air Force Academy; Purdue University            7
Name: count, dtype: int64

The US Naval Academy produced 12 astronauts.

### What subject did the most astronauts major in at college?

In [16]:
df['Undergraduate Major'].value_counts().head(3)

Undergraduate Major
Physics                   35
Aerospace Engineering     33
Mechanical Engineering    30
Name: count, dtype: int64

The same here: use value_counts method on the Undergraduate Major column.
    
The answer is Physics: 35 Astronauts studied physics in college

### Have most astronauts served in the military?

In [17]:
astronauts_with_military_rank = df['Military Rank'].count()
astronauts_with_military_rank

np.int64(207)

The count method returns the number of entries which are not null or not NaN.

207 astronauts have a military rank.

In [18]:
percentage_astronauts_served = astronauts_with_military_rank / len(df)
percentage_astronauts_served

np.float64(0.5798319327731093)

 58% served in the military.

### Which is the most common rank?

In [19]:
df['Military Rank'].value_counts()

Military Rank
Colonel                  94
Captain                  68
Commander                13
Lieutenant Colonel       10
Major                     5
Major General             4
Brigadier General         4
Lieutenant Commander      3
Lieutenant General        2
Rear Admiral              2
Chief Warrant Officer     1
Vice Admiral              1
Name: count, dtype: int64

which gives us 94 Colonels.